### ISSUE TITLE
Build fails because of various errors

### ISSUE THREAD

**Issue by dano** [vert-x3/vertx-lang-python#1]
Attempting a Maven build fails with a very long stack trace, most of which look something like this:

`Address.java:27: error: Could not generate element for io.vertx.core.net.SocketAddress: [Error: unable to resolve method: io.vertx.codegen.Helper.getGenericType(java.lang.String) [arglength=1]]`

A few other errors are mixed in, though:

`[ERROR] diagnostic: /home/dan/code/vert.x/vertx-lang-python/target/vertx-core/io/vertx/core/dns/DnsClient.java:36: error: Could not generate element for io.vertx.core.dns.DnsClient: [Error: unable to resolve method: io.vertx.codegen.Helper.isBasicType(java.lang.String) [arglength=1]]`

`[ERROR] diagnostic: /home/dan/code/vert.x/vertx-lang-python/target/vertx-core/io/vertx/core/Vertx.java:87: error: Could not generate model for vertx(io.vertx.core.VertxOptions): type io.vertx.core.VertxOptions is not legal for use for a parameter in code generation
  static Vertx vertx(VertxOptions options) {`

`[ERROR] diagnostic: /home/dan/code/vert.x/vertx-lang-python/target/vertx-core/io/vertx/core/file/FileSystem.java:492: error: Could not generate model for open(java.lang.String,io.vertx.core.file.OpenOptions,io.vertx.core.Handler<io.vertx.core.AsyncResult<io.vertx.core.file.AsyncFile>>): type io.vertx.core.file.OpenOptions is not legal for use for a parameter in code generation`

`[ERROR] diagnostic: /home/dan/code/vert.x/vertx-lang-python/target/vertx-core/io/vertx/core/eventbus/Message.java:93: error: Could not generate model for reply(java.lang.Object,io.vertx.core.eventbus.DeliveryOptions): type io.vertx.core.eventbus.DeliveryOptions is not legal for use for a parameter in code generation
`

It looks like the `getGenericType` and `isBasicType` methods were removed from vertx-codegen back in Decmber, in this commit: https://github.com/vert-x3/vertx-codegen/commit/e8329c4d940b4eebfd8e7671f1d37b2b97c617c7. I'm not sure if some other method should be used intead, or if those methods need to be added back in. The commit message just said it was cleaning up unused methods, so if that was the only reason for removal, adding them back probably makes sense. No idea about the other errors, though.

**Comment by dano** [comment]
Hmm, a closer look at the javascript template (which it appears the Python template is based on) indicates they've replaced the `getGenericType` calls with a different approach. I'll mess around with it some and see if I can get it working based those changes.

**Comment by kuujo** [comment]
Hey. Yep I started on this before a lot of changes were made to the code gen API. It was generating valid Python code when I left it, but I'm not surprised it's broken now. Would love you to look at it. I'll be around for PRs and any questions as well.

**Comment by dano** [comment]
I finally got around to looking at this. I spent a couple of hours on it tonight and got it at least part of the way back to working again. I'm not sure when I'll have time to continue with it, so here's the branch if anyone else wants to pick up where I left off in the meantime: https://github.com/dano/vertx-lang-python/tree/fix_generation.

**Comment by dano** [comment]
Looks like I'm going to be able to dig into this at work, so I'll have lots of time to work on this :)

I've got the code generation working completely now, but I'm running into a variety of runtime errors when actually using the API (though some things work).

@kuujo, What state was the generated code in when you first added it here? I tried the following code:

``` python
from vertx_python.core.vertx import Vertx

def handler(a):
    print("timer fired {}".format(a))

vertx = Vertx.vertx()
print(vertx)
vertx.set_periodic(1000, handler)
```

And ended up with this exception:

```
Succeeded in deploying verticle 
<vertx_python.core.vertx.Vertx object at 0x7f08cafad590>
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    vertx.set_periodic(1000, handler)
  File "/home/cycdev/code/vertx-lang-python/target/classes/vertx_python/core/vertx.py", line 167, in set_periodic
    return self.jvertx.setPeriodic(delay, LongHandler(handler))
  File "/home/cycdev/envs/vertx/local/lib/python2.7/site-packages/py4j/java_gateway.py", line 538, in __call__
    self.target_id, self.name)
  File "/home/cycdev/envs/vertx/local/lib/python2.7/site-packages/py4j/protocol.py", line 304, in get_return_value
    format(target_id, '.', name, value))
py4j.protocol.Py4JError: An error occurred while calling o48.setPeriodic. Trace:
py4j.Py4JException: Invalid interface name: i
    at py4j.Protocol.getPythonProxy(Protocol.java:383)
    at py4j.Protocol.getObject(Protocol.java:288)
    at py4j.commands.AbstractCommand.getArguments(AbstractCommand.java:81)
    at py4j.commands.CallCommand.execute(CallCommand.java:77)
    at py4j.GatewayConnection.run(GatewayConnection.java:207)
    at java.lang.Thread.run(Thread.java:745)


Exception TypeError: "'NoneType' object is not callable" in <function <lambda> at 0x7f08cafee758> ignored
Exception TypeError: "'NoneType' object is not callable" in <function <lambda> at 0x7f08cb064ed8> ignored
Exception TypeError: "'NoneType' object is not callable" in <function <lambda> at 0x7f08caff5230> ignored
Exception TypeError: "'NoneType' object is not callable" in <function <lambda> at 0x7f08cafc5578> ignored
```

Do you know if calling methods that used Handlers worked ok when you were testing? I'm just trying to figure out if there's a regression here or if this just never worked right.

**Comment by kuujo** [comment]
That's great news!

I'm trying to remember the exact state of the code when I left it. I know I was able to generate valid Python code with the codegen annotations as they were at that time, and I know I was able to deploy a Python verticle, but I'm not sure whether or not this is a regression. I feel like it could be a regression, but I could be wrong about that.

I'm pretty busy finishing up various other projects right now, but I'll be around and definitely available for code reviews and hacking some more code on this as well if necessary. You should be able to get it in a usable state pretty quickly though.

On a side note, this may be best left for another day, but it may be something to think about: with Jython 2.7 out, the code generation portion can probably be useful in a shared Jython/CPython repository with verticle factories implemented separately for Jython and CPython. Look at the JS languages for how to go about that type of abstraction (I think they still support various engines). I can probably throw together a Jython factory.

**Comment by dano** [comment]
Ok, I figured it out.  It's definitely always been broken (unless py4j has changed its API?) - the callback classes being generated used `implements = 'io.vertx.core.Handler'`, but it needs to be `implements =  ['io.vertx.core.Handler']`. The required callback server on the Python side also wasn't being started. After fixing those issues, my example above is now functional!

**Comment by dano** [comment]
As I've been testing out the API, I've noticed that the Vert.x 3 style, which heavy on lambda functions, doesn't map particularly nicely into Python. Consider this Java example:

``` java
public class Send extends AbstractVerticle {

  @Override
  public void start() throws Exception {
    SharedData sd = vertx.sharedData();
    sd.<String, String>getClusterWideMap("mymap", res -> {
      if (res.succeeded()) {
        AsyncMap<String, String> map = res.result();
        System.out.print("Putting 'foo->bar' into the map");
        map.put("foo", "bar", resPut -> {
          if (resPut.succeeded()) {
            System.out.println("Put succeeded.");
          } else {
            System.out.println("Put failed.");
          }
        });
      } else {
        System.out.println("Getting map failed.");
      }
    });
  }
}
```

A roughly equivalent Python script would look like this:

``` python
from vertx_python.core.vertx import Vertx

def clustered_handler(v, exc=None):
    if v:
        sd = v.shared_data()
        def map_handler(m, exc=None):
            def put_handler(_, exc=None):
                if exc is None:
                    print("Put succeeded")
            if m:
                print("Putting 'foo->bar' into the map.")
                m.put('foo', 'bar', put_handler)
            else:
                print("Getting map failed.")
         sd.get_cluster_wide_map("mymap", map_handler)
     else:
         print("Getting clustered vert.x failed")

Vertx.clustered_vertx(clustered_handler)
```

I find this to be very difficult to read. A more Pythonic way to handle this would be to use coroutines. Recent versions of Python have been adding and improving support for this via the [`asyncio`](https://docs.python.org/3/library/asyncio.html) module. Here's the equivalent code if I plug `asyncio` support into the generated API:

``` python
import asyncio

import vertx_python
vertx_python.use_coroutines = True
from vertx_python.core.vertx import Vertx

@asyncio.coroutine
def main():
    try:
        v = yield from Vertx.clustered_vertx()
        sd = v.shared_data()
        m = yield from sd.get_cluster_wide_map("mymap")
        print("Putting 'foo->bar' into the map.")
        out = yield from m.put('foo', 'bar')
        print('put succeeded! {}'.format(out))
    except Exception as e:
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

I've got a rough version of this working now; I added it to the generated code so that it's actually optional - if the call to `vertx_python.use_coroutines = True` isn't made, the regular callback-based system is used.

Does this seem like something worth continuing to pursue? I think many Python developers will find this sort of API far more usable than the callback-based one. And at least as long as Py4J is being used, it's ok to use the `asyncio` event loop, since we don't block the Vert.x calling thread unless we're inside a Vert.x API call. An alternative approach would likely need to be used for Jython, since we'd actually be running inside the JVM.

**Comment by kuujo** [comment]
I think it's great! I've never been much of a fan of Python for async code. Because there are no typical inline closures, it makes for really unreadable code unless you can fit some expression into a lambda. PySpark has the same problem.

I think it would be great to provide an alternative to callbacks that jump up and down the page, but at the same time we should preserve that option to make it easier to migrate Vert.x 2 code to Vert.x 3 as well.

How hard would it be to abstract the codegen in to a base project and provide the two styles as separate projects? I think that might be more in line with the Vert.x model. Also, if we share the codegen with a Jython implementation, would this be compatible Jython?

**Comment by dano** [comment]
@kuujo Adding the `asyncio` interoperability is done just by adding a decorator to the appropriate methods, so it shouldn't be too hard to just replace the bit of code that generates method definitions in one of the child templates, and have everything else live in the parent.

I think a slightly different approach would need to be used for this to work with Jython. Because the Jython code will actually be running inside the Vert.x JVM (unlike the Python/Py4J code), we can't just let the `asyncio` event run forever, since it's going to conflict with/block the Vert.x event loop. It should still be possible to use coroutines, though. I'd just need to write  "runner" of sorts that plugs into the Vert.x event loop. Once I get that working, it could probably be used for both the Py4J and Jython solutions.

**Comment by dano** [comment]
At this point I've got 79 of the API tests working - is it worth submitting a pull request and getting the work I have so far upstream? There are still some things that are probably broken (handling map types being returned, overloaded functions, basically anything that isn't being unit tested yet), but large chunks of the API should be working ok. I figure it might encourage more contributions if there's more working pieces available.

**Comment by dano** [comment]
Another update: I've now got all but 3 of the API tests working. The ones that still fail are failing because of a shortcoming of Py4J's design - it automatically turns Java short/byte types into ints, which makes it impossible to pass e.g. a List<short> from Python to java. This causes casting errors on the Java side. I think the only way to work around it without Py4J changes is to add a Java shim layer between the Python code and the actual vert.x API that can do the necessary type conversions.

I'm still trying to work out the best way to do that - I'm not all that familiar with Java programming, so I'm not sure what the best option there is - whether I can generate java code or if I can somehow pass method names to the java side and have it call the actual methods using reflection. I've also sent mail to the Py4J developer about the issue, and he said it may be possible to turn off auto-unboxing in Py4J, so that a `java.lang.Short(3)` created in Python won't be automatically turned into a Python int. That would probably be the ideal solution, but it's not certain that he'll actually be able to add support for that.

I'm still wondering if a pull request is desirable at this point, too.

**Comment by kuujo** [comment]
That's awesome it's so close! Tons of help.

Yeah we're definitely going to need a shim. I should get back in and play with it. I may be able to help on that side.

Disabling auto unboxing would definitely be ideal, but might take too long to get.

You can open a PR, but we'll have to get those last few tests passing to merge it. That's fine, though, as there will probably be a bit of work to do anyways. Open a PR and I'll try to dig in to the auto boxing issue.

Out of curiosity, do you have any sense of the performance overhead from Py4J?

**Comment by purplefox** [comment]
(Sorry, coming late to the party ;) )- Cool, so this is using Py4J now, not Jython?

**Comment by dano** [comment]
@purplefox Correct. We could certainly use this template as the basis for a Jython template later on, though.

@kuujo The Py4J author got back to me saying he has a plan to prevent primitive wrappers (e.g. java.lang.Short) from being auto-converted to Python ints, but it's non-trivial and would take a few weeks to implement. If you prefer going the shim route rather than waiting on that (assuming the "couple of weeks" estimate is even accurate), that's fine with me.

I haven't done any kind of formal performance testing, but all 121 working unit tests, which all need to call into the JVM, execute in less than a second on the two different machines I've tried (one was around 0.5 seconds, the other 0.8). If you include the overhead for starting the GatewayServer and the Python process, it adds an extra 0.5 seconds or so. Note that all those tests are running in the same Python process and sharing the same py4j JavaGateway/GatewayServer. Startup and especially shutdown time for the Python verticle is (unsurprisingly) really bad; running all 121 unit tests with each test getting a unique Python process/GatewayServer/JavaGateway balloons the run time from under one second to over 500(!) seconds.  the startup time of the Python process and connecting to the GatewayServer is non-trivial, and correctly shutting down the JavaGateway tends to be _very_ slow. It's also very important that the shutdown code always runs, since failing to do so tends to make the Python process hang forever, because py4j runs a callback server in a non-daemon background thread. Without explicitly killing it prior to exiting the main thread, it will keep the process alive forever.

**Comment by ncoghlan** [comment]
(another one late to the party) I was recently curious about the absence of Python examples on the vertx.org home page, and @purplefox directed me here. Thanks for working on this!

Definitely a big +1 for offering a coroutine compatible API on the Python side. For Python 3.5 we're adding dedicated async/await syntax, which makes them read almost as cleanly as the corresponding blocking code: http://www.curiousefficiency.org/posts/2015/07/asyncio-tcp-echo-server.html

Victor Stinner's trollius offers coroutine-like syntax using generators in Python 2, so as long as those can be made to work with vert.x in addition to asyncio coroutines, then the pluggable nature of the asyncio/trollius event loop may make it possible to integrate a Jython based version more directly into the vert.x event loop later (to avoid an https://xkcd.com/927/ type situation, asyncio was designed to allow Twisted, Tornado and GUI frameworks to replace the default event loop with their own, so I suspect the same benefit may apply to vert.x)

**Comment by dano** [comment]
@ncoghlan I've been following PEP 492 very closely, and it was definitely on my mind when I suggested asyncio integration. I also think I may have been the one who piqued your curiosity about the lack of Python support in Vert.x 3 over in the "solving multi-core Python" thread on python-ideas. :)

re: Jython asyncio integration, I haven't done any kind of in depth investigation, but I'm not sure how feasible it will be to _completely_ replace the default `asyncio` event loop with a Vert.x event loop-based implementation, just because Vert.x seems to hide the existence of the event loop from its users as much as possible. Because of that, I'm not sure if we'll have the APIs available on the Vert.x side required to support all the APIs provided by the asyncio loop. At the very least, some kind of partial support should be possible. I also don't see any reason trollius can't be supported, though I haven't tested it with what I've hacked together so far.

In any case, the Py4J implementation can get away with running the default asyncio event loop along side the Vert.x loop, since they're running in separate processes, so that will work as a starting point.

**Comment by kuujo** [comment]
Cool @ncoghlan thanks for the input! I think it would be really nice to have too. It's definitely a huge step in the right direction for async frameworks in a language that can otherwise make for some really unreadable async code.

We'll have to take a look at how the event loop integration works. It may indeed not really benefit this project, but I suppose that depends on the level of abstraction. But coroutines will be a huge win either way.

@purplefox this CPython implementation should resolve a lot of compatibility issues with using Vert.x in Python. I do think perhaps the code generation should be made into a generic base project like the JS implementations were now that Jython 2.7 has been released.

**Comment by purplefox** [comment]
Btw @kuujo did you see my private email I sent you yesterday (kind of) related to this?

**Comment by dano** [comment]
@ncoghlan I gave trollius a spin in Jython, and it turns out it doesn't work. It looks like there's probably a fair amount of work needed to fix it - trollius is using a bunch of modules/methods that don't exist on Jython (`socket.socketpair`, `fcntl`, etc.).

**Comment by kuujo** [comment]
Still digging through them let me just scroll down :-)

**Comment by kuujo** [comment]
@dano in that case, I think if we are going to have a Jython implementation (which has performance benefits and I think at least some people want it) the compatibility issues may dictate that we do need a base Python language implementation that uses basic callbacks to be shared by both Jython and CPython. We should stick with what is here ATM and anything like Trollius can be done in a separate project whether that's inside or outside of the Vert.x organization.

**Comment by dano** [comment]
@kuujo That's fair. Another possibility I was thinking about was something along the lines of vertx-sync, but for Python/Jython - basically, suspendable coroutines very much along the lines of what `asyncio` gives you, but just using native Python coroutine features and the Vert.x event loop, rather relying on the `asyncio` event loop. Any method that takes a `Handler<AsyncResult<X>>` as its final argument (just like vertx-sync) would return a `Future` of some kind, rather than None. Unlike vertx-sync, it could still optionally take a callback. So the Python code would look something like this (I'm leaving out parameter validation, type conversion, etc. here, but you get the idea):

``` python

def clustered_vertx(self, result_handler=None, **options):
    fut = Future()  # This is very similar to `asyncio.Future`, but wouldn't need the asyncio event loop.
    _handler = wrap_handler(result_handler, fut)
    util.jvm.io.vertx.core.Vertx.clusteredVertx(VertxOptions(options), _handler)
    return fut
```

`wrap_handler` would look like this:

``` python
def wrap_handler(user_handler, fut):
    def set_result(result, exc):
        if exc is not None:
             fut.set_exception(exc)
        else:
             fut.set_result(result)
        if user_handler is not None:
             user_handler(result, exc)
```

A user could optionally wait on the future that's returned using `yield`/`yield from`, or just let the callback they passed fire:

``` python
def on_got_cv(v):
    print("Got {}".format(v))

if __name__ == "__main__":
    clustered_vertx(on_got_cv)  # Still returns a future, but you can ignore it.
```

Or

``` python
@coroutine
def main():
    v = yield from clustered_vertx()  # You could still pass a callback here.
    print("Got {}".format(v))

if __name__ == "__main__":
    vertx_runner.run(main())  # Uses the vertx event loop to drive the app.
```

I haven't actually coded this up, but I _think_ something like it should be possible for both Py4J and Jython. If it proves to be doable, it could theoretically be the default API for both Python and Jython, since it doesn't prevent a purely callback based approach. Of course, it could also just be done as a separate project, same as vertx-sync.

**Comment by kuujo** [comment]
I think that's awesome and totally possible! We'll have to discuss this on the dev group since it diverges from the Vert.x API, though obviously not much since the future is an additional return value. But I think it's a very interesting proposition.

What I'd be concerned about is confusion. We're essentially providing two ways of accomplishing the same task, but we have to make sure users can't use both. I think we'd have to ensure if a typical async handler is passed to a method, None is returned, and that's fine.

There's some benefit to the async handler approach in comparison to futures, and it may be an issue. Specifically, in many cases Vert.x optimizes cases where no handler is passed to an overloaded method. For instance, if no reply handler is passing to the even bus's send() method, Vert.x doesn't need to store the handler and context for the reply. Assuming that in all cases the user might want some feedback on when an asynchronous operation is complete could dampen performance and create unnecessary garbage in the JVM. There are certainly ways around that, but it really depends on how clean the solutions are.

Let me collect my thoughts on this tonight and bring it up in the dev group where hopefully more people can contribute to the decision making process.

**Comment by dano** [comment]
> What I'd be concerned about is confusion. We're essentially providing two ways of accomplishing the same task, but we have to make sure users can't use both. I think we'd have to ensure if a typical async handler is passed to a method, None is returned, and that's fine.

That makes sense. I implemented something like this over the weekend in the [python_sync branch](https://github.com/dano/vertx-lang-python/tree/python_sync) of my fork. It works with both Python 2.x and Python 3.3+, though the exact usage varies a bit from version to version (because of changes in the language's coroutine support). Python 2.x looks like this:

``` python
from vertx_python.futures import coroutine, VertxRunner, Return
from vertx_python.core.vertx import Vertx
from vertx_python.util import handle_vertx_shutdown

@coroutine
def test():
    v = yield Vertx.clustered_vertx()   # yield instead of yield from; yield from is not supported in Python 2.x.
    sd = v.shared_data()
    m = yield sd.get_cluster_wide_map('mymap')
    yield m.put('foo', 'bar')
    out = yield m.get('foo')
    raise Return(out)   # In Python 2.x, generators can't return anything, so this workaround is used.

@coroutine
def test2():
    v = yield test()
    raise Return(v)

if __name__ == "__main__":
    with handle_vertx_shutdown():
        out = VertxRunner().run_until_complete(test2())
        print("got out {}".format(out))
```

While the equivalent Python 3.5 version looks like this:

``` python
from vertx_python.futures import VertxRunner
from vertx_python.core.vertx import Vertx
from vertx_python.util import handle_vertx_shutdown

async def test():  # No decorator needed thanks to async/await support in Python 3.5.
    v = await Vertx.clustered_vertx()   # await instead of yield/yield from in Python 3.5.
    sd = v.shared_data()
    m = await sd.get_cluster_wide_map('mymap')
    await m.put('foo', 'bar')
    out = await m.get('foo')
    return out

async def test2():
    v = await test()
    return v

if __name__ == "__main__":
    with handle_vertx_shutdown():
        out = VertxRunner().run_until_complete(test2())
        print("got out {}".format(out))
```

In both cases, the Python API uses a keyword argument for `Handler<AsyncResult<X>>` whenever it appears as the last argument to a Java function. If the keyword argument is provided by the user, no `Future` is returned, the handler is passed as-is to the JVM, and behavior is exactly as it is now in the `fix_generation` branch. The `Future`/`coroutine` stuff only kicks in if the user doesn't provide a handler.

Note that this doesn't provide any `asyncio` interoperability - it just provides the plumbing required to use coroutines with the Vert.x event loop. It uses `concurrent.futures.Future` (which is built-in on Python 3.x, and available as a backport for 2.x) for its `Future` implementation, and a coroutine engine based on the one provided by `tornado`. Some internal modifications will likely be needed to make it work with Jython, but they shouldn't be too extensive, and I believe the user-facing API can be kept the same.

Hopefully having a reference implementation that should work with both pure callbacks and coroutines for CPython 2.x/3.x and Jython will help guide the dev group discussion.

**Comment by hegdeaditya** [comment]
Hi guys, I went through the discussion above but didn't quiet clearly understand if there is an official support for python in Vertx-3. Can anyone tell me the current status of python support? Thanks in advance!

**Comment by vietj** [comment]
@hegdeaditya it is supposed to, but nobody is contributing to this effort

**Comment by dano** [comment]
@hegdeaditya I had the build working back in May in this branch: https://github.com/dano/vertx-lang-python/tree/fix_generation. Not sure if it will work now.

However, I ended up shifting my focus to integrating with Jython, rather than CPython, because I felt like that was a better starting point. The CPython -> JavaGateway required when using Py4J introduces a some complications that were getting in the way. And now that Jython has released 2.7 compatibility, it's a much more attractive option than it was when the Py4J work started.

I have the Jython work here, but I haven't touched it since July: https://github.com/dano/vertx-lang-jython. I honestly can't remember where exactly I left it. I think it builds, but a bunch of the unit tests fail. The last thing I remember doing was trying to figure out how to run Jython unit tests directly via Maven.

I have interest in continuing to work on it, but I've just been too busy to spend the time.

**Comment by vietj** [comment]
@dano would you like to contribute it to Vert.x instead of this cpython impl ?

**Comment by dano** [comment]
@vietj That's the plan, if I ever find the time to continue working on it :)

**Comment by Danilo-Araujo-Silva** [comment]
I would like to thank all of this effort and ask please to add python support ^^

Since python is a game change language for some ML operations, have a first class python support would be really awesome!
